File: apps/backend/Python-backend/ModuAgent/core/interfaces/memory.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional


class BaseMemory(ABC):
    @abstractmethod
    def query(
        self,
        user_id: str,
        context_window: str,
        required_fields: List[str],
    ) -> Dict[str, Any]:
        pass

    @abstractmethod
    def update(
        self,
        user_id: str,
        new_data: Dict[str, Any],
        metadata: Dict[str, Any],
    ) -> bool:
        pass


class BaseStorageAdapter(ABC):
    @abstractmethod
    def adapter_type(self) -> str:
        pass

    @abstractmethod
    def load(self, key: str) -> Optional[Dict[str, Any]]:
        pass

    @abstractmethod
    def save(self, key: str, data: Dict[str, Any]) -> bool:
        pass

File: packages/modu-agent/src/core/interfaces/memory.ts
// 对应 Python: core/interfaces/memory.py
// BaseMemory + BaseStorageAdapter 抽象接口

/**
 * 记忆抽象接口。
 * 对应 Python BaseMemory（query / update）。
 */
export abstract class BaseMemory {
  abstract query(
    userId: string,
    contextWindow: string,
    requiredFields: string[],
  ): Promise<Record<string, any>> | Record<string, any>

  abstract update(
    userId: string,
    newData: Record<string, any>,
    metadata: Record<string, any>,
  ): Promise<boolean> | boolean
}

/**
 * 存储适配器抽象接口。
 * 对应 Python BaseStorageAdapter（adapter_type / load / save）。
 */
export abstract class BaseStorageAdapter {
  abstract adapterType(): string

  abstract load(key: string): Promise<Record<string, any> | null> | Record<string, any> | null

  abstract save(key: string, data: Record<string, any>): Promise<boolean> | boolean
}
